### PR TITLE
Fix crash for passwords starting with '@' for Issue #7707

### LINKED ISSUE #7707: Nuget.exe sources add with -password "@foo" raises System.FormatException
## Details about Problem

NuGet product used: NuGet.exe

NuGet version: 4.9.2.5706

OS version (i.e. win10 v1607 (14393.321)): win10 v1709 16299.64

Worked before? If so, with which NuGet version:

Not to my knowledge

## Detailed repro steps so we can see the same problem

1. Open Command Line

2. Enter Command to add a new nuget source. e.g.

    nuget sources add -name foo.bar -source C:\NuGet\local -username foo -password "@Blah"

   Important: The first char of the password should be a '@'

3. The following Exception occures: 
```
Unhandled Exception: System.FormatException: Index (zero based) must be greater than or equal to zero and less than the size of the argument list.

   at System.Text.StringBuilder.AppendFormatHelper(IFormatProvider provider, String format, ParamsArray args)

   at System.String.FormatHelper(IFormatProvider provider, String format, ParamsArray args)

   at System.String.Format(String format, Object[] args)

   at NuGet.Common.CommandLineResponseFile.ParseArgsResponseFiles(String[] args, Int32 parseArgsResponseFileRecursionDepth)

   at NuGet.CommandLine.Program.Main(String[] args)
```
### Verbose Logs

Verbose log was the same as the Exception provided under Step 3

### Sample Project

The issue does not apply to a certain project.

### PR DESCRIPTION
**Pull request by brann0n**
## Bug

Fixes: [NuGet/Home#7707](https://github.com/NuGet/Home/issues/7707)
Regression: Yes/No  
* Last working version:  Unsure when this stopped working, i only started using nuget.exe today
proof
## Fix

Details: In the check for arguments starting with an '@' i added a check for password so that it doesnt crash when a password that starts with '@' is used. 
A check for other CLI arguments might also be needed, but i discovered the issue upon entering my password.

## Testing/Validation

Tests Added: No
Reason for not adding tests:  I couldnt get the project to compile, it errord on the 'extern alias CoreV2;' in Commands.cs and  several other files.

### CODE REVIEW

**Comment by zivkan** [comment]
@brann0n thanks for taking the time to create a pull request.

> I couldnt get the project to compile, it errord on the 'extern alias CoreV2;' in Commands.cs and several other files.

As per the [build instructions in our contribution guide](https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#build), you need to run `configure.ps1`. Specifically you need to restore the project `build\bootstrap.proj`, at least to get the nuget.core assembly, but my memory is that's all that's needed to get the solution to compile.

Looking at the file you modified, I see this comment:
https://github.com/NuGet/NuGet.Client/blob/02ca43109f78ffba944bde0727e19652ee99c0a6/src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineResponseFile.cs#L15

I don't know if we have any tests for this already, but I don't feel comfortable accepting this PR without some tests. Regardless of tests, this PR will break the scenario where a customer has written their password to a file and uses `nuget.exe .... -password @password.txt` to avoid putting the password on the command line (think CI scenarios and build logs leaking secrets). This is something that would probably need to be written in the linked issue, so that if this PR gets merged, customers looking at the release notes can be aware of this breaking change in behaviour. We don't have telemetry from nuget.exe, so we have no way of knowing if this feature is used or not.

**Comment by brann0n** [comment]
@zivkan thanks for answering and helping me out.

I ran the powershell script from the package manager but somehow that messed things up, so I re cloned the project and used proper powershell to make it work.

I also undid my previous changes, and am going to work on a solution that doesn't break existing functionality later this week.

**Comment by ghost** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/NuGet/NuGet.Client?pullRequest=3822) <br/>All CLA requirements met.

**Comment by dominoFire** [comment]
Thanks for your contribution, @brann0n 

There are two failing tests:

- `NuGet.PackageManagement.VisualStudio.Test.NuGetPackageSearchServiceTests.GetTotalCountAsync_WithGreaterThanOrEqualToMaxCountResults_ReturnsMaxCount` at test\NuGet.Clients.Tests\NuGet.PackageManagement.VisualStudio.Test\Services\NuGetPackageSearchServiceTests.cs
- (fails only on MacOS) `RestoreLogging_NetCore_NoWarnWithTreatWarningsAsErrorRemovesWarningAsync` at test\NuGet.Clients.Tests\NuGet.CommandLine.Test\RestoreLoggingTests.cs

Please, take a look at those tests. If you need help, just let us know.

Thanks!

**Comment by brann0n** [comment]
Hello, 
Thank you for commenting.
I do not understand, i checked the code for the tests you mentioned and
my code changes do not seem to change anything in a way that this test should fail...

**Comment by dominoFire** [comment]
@brann0n Thanks for the reply! It seems those tests I mentioned earlier were regressed in another part.

We need a unit test for this change. You can take inspiration from the test linked below and create a new one with your modified parameters in the same `NuGetSourcesCommandTest.cs` file.

https://github.com/NuGet/NuGet.Client/blob/3b5c425bea75ba20cb6136e37944176c9e45c774/test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetSourcesCommandTest.cs#L48-L92

File: src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineResponseFile.cs [hunk 1]
```diff
@@ -64,7 +65,11 @@ private static string[] ParseArgsResponseFiles(string[] args, int parseArgsRespo
 
                 if (!File.Exists(responseFilePath))
                 {
-                    throw new ArgumentException(string.Format(LocalizedResourceManager.GetString("Error_ResponseFileDoesNotExist")), arg);
+                    //if the file could not be found, report this back to the user, but dont crash yet, the use of '@' in the argument might be intentional here.
+                    System.Console.WriteLine(string.Format(LocalizedResourceManager.GetString("Error_ResponseFileDoesNotExist"), arg));
```
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.CommandLine/Common/CommandLineResponseFile.cs:69]: You may need to add additional validation to verify you are not printing the password back to the console

**Comment by dominoFire** [comment]
Please, make sure you’re working on top of the development branch to ease the final merging and to have the latest NuGet changes.

Here’s a script that could help you rebase your current branch with the changes on top of the dev branch. If you need help, let us know.

https://gist.github.com/dominoFire/3fdb4c63c51d6faf264a83bef44b88e2

**Comment by dominoFire** [comment]
@brann0n Friendly ping 🔔. Is everything alright?

**Comment by dominoFire** [comment]
Hi @brann0n 

Hope all is well.

Are you working on this PR? It's been a while since we heard from you. I hope you and your loved ones are doing well in these trying times.

We would love to hear back from you. As stated in our [Contribution guide](https://github.com/NuGet/NuGet.Client/blob/dev/CONTRIBUTING.md#contributing-step-by-step), we will have to close this PR if there's no activity within the next 3 days.

Let us know if you need help.

**Comment by brann0n** [comment]
Hi, thanks for the ping. 
I resigned from the company where fixing this issue would have improved daily tasks.
With my new job i no longer have the time to work on this, although this issue still seems important to fix.

Maybe in the future someone else can pick it up were i left it.

Kind Regards!

**Comment by dominoFire** [comment]
Ok, I will close the PR for now. Thanks for the help, @brann0n !

**Comment by WillMarcouiller** [comment]
Has this been fixed?  I run into the same problem whilst trying to add a nuget source for azure artifacts.  That is quite confusing when you're not used to nuget cli and this happens.  I first thought I didn't do something correctly until I figure out this bug on the net.

**Pull request closed without merging**